File: frontend/src/types/graph.ts
// Graph data types — matches backend app/models/graph.py exactly

export interface GraphNode {
  id: string; // sys_id (32 hex chars)
  name: string;
  ci_class: string; // Normalized label: "Server", "Database", etc.
  ci_class_raw: string; // Original sys_class_name
  environment: string;
  operational_status: number; // 1=Operational, 2=Non-Op, 3=Repair, 4=Retired
  degree: number;
  cluster_id: number;
  x?: number;
  y?: number;
  z?: number;
  // Internal force-sim velocity (set by d3-force-3d at runtime)
  vx?: number;
  vy?: number;
  vz?: number;
  // Internal force-sim fixed positions
  fx?: number;
  fy?: number;
  fz?: number;
}

export interface GraphLink {
  source: string | GraphNode; // sys_id or resolved node ref
  target: string | GraphNode;
  rel_type: string; // "Runs on", "Depends on", etc.
  rel_type_reverse: string;
}

export interface NeighborhoodResponse {
  nodes: GraphNode[];
  edges: GraphLink[];
  center_id: string;
  total_in_neighborhood: number;
  truncated: boolean;
  query_time_ms: number;
  cached: boolean;
}

export interface PathStep {
  path: string[]; // list of sys_ids
  length: number;
  edge_types: string[];
}

export interface PathResponse {
  paths: PathStep[];
  source_id: string;
  target_id: string;
  exists: boolean;
  query_time_ms: number;
  cached: boolean;
}

export interface ClassCount {
  ci_class: string;
  count: number;
}

export interface TypeCount {
  rel_type: string;
  count: number;
}

export interface EnvCount {
  environment: string;
  count: number;
}

export interface CISample {
  sys_id: string;
  name: string;
  ci_class: string;
}

export interface ClusterSummary {
  cluster_id: number;
  size: number;
  label: string;
  top_ci_classes: ClassCount[];
  sample_cis: CISample[];
}

export interface ClustersResponse {
  clusters: ClusterSummary[];
  total_clusters: number;
  query_time_ms: number;
  cached: boolean;
}

export interface GraphStatsResponse {
  total_nodes: number;
  total_edges: number;
  avg_degree: number;
  max_degree: number;
  ci_class_distribution: ClassCount[];
  rel_type_distribution: TypeCount[];
  env_distribution: EnvCount[];
  query_time_ms: number;
  cached: boolean;
}

export interface NeighborhoodOptions {
  hops?: 1 | 2 | 3;
  maxNodes?: number;
  degreeThreshold?: number;
  classFilter?: string[];
  envFilter?: string[];
}

export interface NodeConfig {
  shape: "sphere" | "box" | "octahedron" | "cylinder" | "cone" | "torus" | "dodecahedron" | "icosahedron";
  color: string;
  size: number;
  glowIntensity: number;
}

export interface EdgeStyle {
  color: string;
  dashed: boolean;
  width: number;
}

File: backend/app/models/graph.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from app.models.base import SysId, TimestampedMixin


class ClassCount(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    ci_class: str
    count: int


class TypeCount(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    rel_type: str
    count: int


class EnvCount(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    environment: str
    count: int


class GraphNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: SysId
    name: str
    ci_class: str
    ci_class_raw: str
    environment: str
    operational_status: int
    degree: int = 0
    cluster_id: int = -1
    x: float | None = None
    y: float | None = None
    z: float | None = None


class GraphEdge(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    source: SysId
    target: SysId
    rel_type: str
    rel_type_reverse: str = ""


class NeighborhoodResponse(TimestampedMixin):
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    center_id: SysId
    total_in_neighborhood: int
    truncated: bool = False


class PathStep(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    path: list[SysId]
    length: int
    edge_types: list[str] = Field(default_factory=list)


class PathResponse(TimestampedMixin):
    paths: list[PathStep]
    source_id: SysId
    target_id: SysId
    exists: bool


class CISample(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    sys_id: SysId
    name: str
    ci_class: str


class ClusterSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    cluster_id: int
    size: int
    label: str = ""
    top_ci_classes: list[ClassCount] = Field(default_factory=list)
    sample_cis: list[CISample] = Field(default_factory=list)


class ClustersResponse(TimestampedMixin):
    clusters: list[ClusterSummary]
    total_clusters: int


class GraphStatsResponse(TimestampedMixin):
    total_nodes: int
    total_edges: int
    avg_degree: float
    max_degree: int
    ci_class_distribution: list[ClassCount] = Field(default_factory=list)
    rel_type_distribution: list[TypeCount] = Field(default_factory=list)
    env_distribution: list[EnvCount] = Field(default_factory=list)
